packages/resource-deleter/src/main.js

```diff
@@ -0,0 +1,185 @@
+/* @flow */
+import fetch from 'node-fetch'
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import {
+  createAuthMiddlewareWithExistingToken,
+  createAuthMiddlewareForClientCredentialsFlow,
+} from '@commercetools/sdk-middleware-auth'
+import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import { createQueueMiddleware } from '@commercetools/sdk-middleware-queue'
+import type {
+  ApiConfigOptions,
+  LoggerOptions,
+  resourceDeleterOptions,
+} from 'types/resourceDeleter'
+import type {
+  Client,
+  ClientRequest,
+  ClientResponse,
+  MethodType,
+} from 'types/sdk'
+import silentLogger from './utils/silent-logger'
+import pkg from '../package.json'
+
+export default class ResourceDeleter {
+  // Setting type annotations
+  apiConfig: ApiConfigOptions
+  client: Client
+  logger: LoggerOptions
+  predicate: ?string
+  resource: string
+
+  constructor(options: resourceDeleterOptions) {
+    if (!options.apiConfig)
+      throw new Error('The constructor must passed an `apiConfig` object')
+
+    if (!options.resource) throw new Error('A `resource` object must be passed')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          options.accessToken ? `Bearer ${options.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow({
+          ...this.apiConfig,
+          fetch,
+        }),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl, fetch }),
+        createQueueMiddleware({
+          concurrency: 20,
+        }),
+      ],
+    })
+    this.logger = {
+      ...silentLogger,
+      ...options.logger,
+    }
+    this.predicate = options.predicate
+    this.resource = options.resource
+  }
+
+  run(): Promise<any> {
+    this.logger.info(`Starting to delete fetched ${this.resource}`)
+    const uri = this.buildUri(this.apiConfig.projectKey, this.predicate)
+    const request = ResourceDeleter.buildRequest(uri, 'GET')
+
+    return this.client
+      .process(
+        request,
+        (response: ClientResponse): Promise<any> => {
+          if (response.statusCode !== 200) {
+            return Promise.reject(
+              new Error(`Request returned status code ${response.statusCode}`)
+            )
+          }
+          let results = response.body?.results
+
+          // Check if the resource is empty
+          if (!results || !results.length) {
```

Comment:
Hello @Babanila  .. the `client.process()` method calls the processing function for every batch of items loaded from API. If there are no results loaded, it won't call processing function which means that this code should not be needed at all.

Reply:
so based on the internal parts of process method it calls the callback function every time at least once (even if there are no results)